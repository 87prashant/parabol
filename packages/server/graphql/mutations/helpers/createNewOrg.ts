import getRethink from '../../../database/rethinkDriver'
import shortid from 'shortid'
import {OrgUserRole, TierEnum} from 'parabol-client/types/graphql'

export default async function createNewOrg (orgId, orgName, leaderUserId) {
  const r = getRethink()
  const now = new Date()
  return r({
    org: r.table('Organization').insert({
      id: orgId,
      creditCard: {},
      createdAt: now,
      name: orgName,
      tier: TierEnum.personal,
      updatedAt: now
    }),
    organizationUser: r.table('OrganizationUser').insert({
      id: shortid.generate(),
      inactive: false,
      joinedAt: now,
      newUserUntil: now,
      orgId,
      removedAt: null,
      role: OrgUserRole.BILLING_LEADER,
      userId: leaderUserId
    })
  })
}
