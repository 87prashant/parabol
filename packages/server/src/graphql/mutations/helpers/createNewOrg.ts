import getRethink from '../../../database/rethinkDriver'
import {OrgUserRole, TierEnum} from 'parabol-client/src/types/graphql'
import Organization from '../../../database/types/Organization'
import OrganizationUser from '../../../database/types/OrganizationUser'

export default async function createNewOrg(orgId: string, orgName: string, leaderUserId: string) {
  const r = await getRethink()
  const org = new Organization({id: orgId, tier: TierEnum.personal, name: orgName})
  const orgUser = new OrganizationUser({
    orgId,
    userId: leaderUserId,
    role: OrgUserRole.BILLING_LEADER
  })
  return r({
    org: r.table('Organization').insert(org),
    organizationUser: r.table('OrganizationUser').insert(orgUser)
  }).run()
}
