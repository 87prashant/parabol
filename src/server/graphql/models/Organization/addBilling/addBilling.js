import {GraphQLBoolean, GraphQLID, GraphQLNonNull} from 'graphql';
import stripe from 'server/billing/stripe';
import getRethink from 'server/database/rethinkDriver';
import makeUpcomingInvoice from 'server/graphql/models/Invoice/makeUpcomingInvoice';
import getCCFromCustomer from 'server/graphql/models/Organization/addBilling/getCCFromCustomer';
import {getUserId, getUserOrgDoc, requireOrgLeader, requireWebsocket} from 'server/utils/authorization';
import {fromEpochSeconds, toEpochSeconds} from 'server/utils/epochTime';
import getPubSub from 'server/utils/getPubSub';
import sendSegmentEvent from 'server/utils/sendSegmentEvent';
import {ACTION_MONTHLY, TRIAL_EXTENSION, TRIAL_PERIOD} from 'server/utils/serverConstants';
import {NOTIFICATIONS_CLEARED, PAYMENT_REJECTED, TRIAL_EXPIRED, TRIAL_EXPIRES_SOON} from 'universal/utils/constants';

export default {
  type: GraphQLBoolean,
  description: `Add a credit card by passing in a stripe token encoded with all the billing details.
  Handles 4 scenarios:
    1) Updating to a new credit card
    2) Adding to extend the free trial
    3) Converting after the trial ended
    4) Payment was rejected`,
  args: {
    orgId: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'the org requesting the changed billing'
    },
    stripeToken: {
      type: new GraphQLNonNull(GraphQLID),
      description: 'The token that came back from stripe'
    }
  },
  async resolve(source, {orgId, stripeToken}, {authToken, socket}) {
    const r = getRethink();
    const now = new Date();

    // AUTH
    requireWebsocket(socket);
    const userId = getUserId(authToken);
    const userOrgDoc = await getUserOrgDoc(userId, orgId);
    requireOrgLeader(userOrgDoc);

    // RESOLUTION
    const {creditCard, stripeId, stripeSubscriptionId, periodEnd, periodStart, orgUsers} = await r.table('Organization')
      .get(orgId)
      .pluck('creditCard', 'orgUsers', 'periodEnd', 'periodStart', 'stripeId', 'stripeSubscriptionId');
    const customer = await stripe.customers.update(stripeId, {source: stripeToken});
    if (periodEnd > now && stripeSubscriptionId) {
      // 1) Updating to a new credit card
      if (creditCard.last4) {
        await r.table('Organization').get(orgId).update({
          creditCard: getCCFromCustomer(customer)
        });
        sendSegmentEvent('addBilling Update Payment Success', userId, {orgId});
        // 2) Adding to extend the free trial
      } else {
        const extendedPeriodEnd = new Date(periodStart.setMilliseconds(0) + TRIAL_PERIOD + TRIAL_EXTENSION);
        stripe.subscriptions.update(stripeSubscriptionId, {
          trial_end: toEpochSeconds(extendedPeriodEnd)
        });
        const {removedNotification} = await r({
          orgUpdated: r.table('Organization').get(orgId).update({
            creditCard: getCCFromCustomer(customer),
            periodEnd: extendedPeriodEnd
          }),
          removedNotification: r.table('Notification')
            .getAll(orgId, {index: 'orgId'})
            .filter({
              type: TRIAL_EXPIRES_SOON
            })
            .delete({returnChanges: true})('changes')(0)('old_val').pluck('id', 'userIds').default(null)
        });
        if (removedNotification) {
          const notificationsCleared = {deletedIds: [removedNotification.id]};
          removedNotification.userIds.forEach((notifiedUserId) => {
            getPubSub().publish(`${NOTIFICATIONS_CLEARED}.${notifiedUserId}`, {notificationsCleared});
          });
        }
        sendSegmentEvent('addBilling Free Trial Extended', userId, {orgId});
      }
    } else {
      // 3) Converting after the trial ended
      // 4) Payment was rejected and they're adding a new source
      const notificationToClear = creditCard.last4 ? PAYMENT_REJECTED : TRIAL_EXPIRED;
      const quantity = orgUsers.reduce((count, orgUser) => orgUser.inactive ? count : count + 1, 0);
      const subscription = await stripe.subscriptions.create({
        customer: stripeId,
        metadata: {
          orgId
        },
        plan: ACTION_MONTHLY,
        quantity
      });
      const {id, current_period_end, current_period_start} = subscription;
      const {removedNotification} = await r({
        updatedOrg: r.table('Organization').get(orgId).update({
          creditCard: getCCFromCustomer(customer),
          periodEnd: fromEpochSeconds(current_period_end),
          periodStart: fromEpochSeconds(current_period_start),
          stripeSubscriptionId: id
        }),
        updatedTeam: r.table('Team')
          .getAll(orgId, {index: 'orgId'})
          .update({
            isPaid: true
          }),
        removedNotification: r.table('Notification')
          .getAll(orgId, {index: 'orgId'})
          .filter({
            type: notificationToClear
          })
          .delete({returnChanges: true})('changes')(0)('old_val').pluck('id', 'userIds').default(null)
      });
      if (removedNotification) {
        const notificationsCleared = {deletedIds: [removedNotification.id]};
        removedNotification.userIds.forEach((notifiedUserId) => {
          getPubSub().publish(`${NOTIFICATIONS_CLEARED}.${notifiedUserId}`, {notificationsCleared});
        });
      }
      sendSegmentEvent('addBilling New Payment Success', userId, {orgId, quantity});
    }
    // nuke the upcoming invoice if it existed
    await r.table('Invoice')
      .get(`upcoming_${orgId}`)
      .delete();

    const channel = `upcomingInvoice/${orgId}`;
    const upcomingInvoice = await makeUpcomingInvoice(orgId, stripeId, stripeSubscriptionId);
    const payload = {
      type: 'update',
      fields: upcomingInvoice
    };
    socket.emit(channel, payload);
    return true;
  }
};
