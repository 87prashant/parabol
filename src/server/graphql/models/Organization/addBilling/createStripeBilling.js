import getRethink from 'server/database/rethinkDriver';
import stripe from 'server/billing/stripe';

export default async function createStripeBilling(orgId, stripeToken, validUntil) {
  const r = getRethink();
  const stripeId = await r.table('Organization').get(orgId)('stripeId').default(null);
  const customer = await stripe.customers.update(stripeId, {source: stripeToken});
  const card = customer.sources.data.find((source) => source.id === customer.default_source);
  const {brand, last4, exp_month: expMonth, exp_year: expYear} = card;
  const expiry = `${expMonth}/${String(expYear).substr(2)}`;
  const updateObject = {
    creditCard: {
      brand,
      last4,
      expiry
    }
  };
  if (validUntil !== undefined) {
    updateObject.validUntil = validUntil;
  }
  return r.table('Organization').get(orgId).update(updateObject);
}
