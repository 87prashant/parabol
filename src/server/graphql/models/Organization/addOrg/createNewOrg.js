import getRethink from 'server/database/rethinkDriver';
import stripe from 'server/billing/stripe';
import {ACTION_MONTHLY, TRIAL_PERIOD_DAYS} from 'server/utils/serverConstants';
import {fromStripeDate} from 'server/billing/stripeDate';
import {BILLING_LEADER} from 'universal/utils/constants';
import updateStripeBilling from 'server/graphql/models/Organization/addBilling/updateStripeBilling';

export default async function createNewOrg(orgId, orgName, leaderUserId, stripeToken) {
  const r = getRethink();
  const now = new Date();

  const {id: stripeId} = await stripe.customers.create({
    metadata: {
      orgId
    }
  });
  const {id: stripeSubscriptionId, current_period_end} = await stripe.subscriptions.create({
    customer: stripeId,
    metadata: {
      orgId
    },
    plan: ACTION_MONTHLY,
    // if a payment token is provided, this isn't a trial
    trial_period_days: stripeToken ? 0 : TRIAL_PERIOD_DAYS
  });
  const validUntil = fromStripeDate(current_period_end);
  const creditCard = stripeToken && await updateStripeBilling(orgId, stripeId, stripeToken);

  return await r.table('Organization').insert({
    id: orgId,
    creditCard,
    createdAt: now,
    isTrial: true,
    name: orgName,
    orgUsers: [{ id: leaderUserId, role: BILLING_LEADER}],
    stripeId,
    stripeSubscriptionId,
    updatedAt: now,
    validUntil
  }, {returnChanges: true})('changes')(0)('new_val');
}
