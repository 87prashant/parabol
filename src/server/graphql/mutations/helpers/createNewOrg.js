import getRethink from 'server/database/rethinkDriver'
import {BILLING_LEADER, PERSONAL} from 'universal/utils/constants'
import shortid from 'shortid'

export default async function createNewOrg(orgId, orgName, leaderUserId) {
  const r = getRethink()
  const now = new Date()
  return r({
    org: r.table('Organization').insert({
      id: orgId,
      creditCard: {},
      createdAt: now,
      name: orgName,
      tier: PERSONAL,
      updatedAt: now
    }),
    organizationUser: r.table('OrganizationUser').insert({
      id: shortid.generate(),
      inactive: false,
      joinedAt: now,
      newUserUntil: now,
      orgId,
      removedAt: null,
      role: BILLING_LEADER,
      userId: leaderUserId
    })
  })
}
