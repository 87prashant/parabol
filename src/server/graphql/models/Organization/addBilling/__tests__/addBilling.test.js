import getRethink from 'server/database/rethinkDriver';
import createNewOrg from 'server/__tests__/setup/createNewOrg';
import addBilling from 'server/graphql/models/Organization/addBilling/addBilling';
import mockAuthToken from 'server/__tests__/setup/mockAuthToken';
import stripe from 'server/billing/stripe';
import MockDate from 'mockdate';
import mockNow from 'server/__tests__/setup/mockNow';
import fetchAndTrim from 'server/__tests__/utils/fetchAndTrim';

MockDate.set(mockNow);

let docs;
beforeEach(async () => {
  docs = await createNewOrg();
  stripe.__setMockData(docs.org);
});
console.error = jest.fn();

test('add billing to existing org in a trial period', async () => {
  const r = getRethink();
  const {users, org} = docs;
  const stripeToken = 'tok_4242424242424242';
  const orgId = org.id;
  const authToken = mockAuthToken(users[0]);
  const socket = {};
  await addBilling.resolve(undefined, {orgId, stripeToken}, {authToken, socket});
  const db = await fetchAndTrim({
    organization: r.table('Organization').get(orgId),
    notification: r.table('Notification').getAll(orgId, {index: 'orgId'})
  });
  expect(db).toMatchSnapshot();
  expect(stripe.customers.__snapshot()).toMatchSnapshot();
  expect(stripe.subscriptions.__snapshot()).toMatchSnapshot();
});
