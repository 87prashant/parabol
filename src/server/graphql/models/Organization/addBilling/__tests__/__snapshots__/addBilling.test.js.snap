// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addBilling changes cards for customer who had a failed charge 1`] = `
Object {
  "notification": Array [],
  "organization": Object {
    "createdAt": 2017-02-26T05:42:20.836Z,
    "creditCard": Object {
      "brand": "Visa",
      "expiry": "12/23",
      "id": "card_failCharge",
      "last4": "4242",
      "number": "4000000000000341",
    },
    "id": 0,
    "name": "The Averagers, Inc.",
    "orgUsers": Array [
      Object {
        "id": 2,
        "inactive": false,
        "role": "billingLeader",
      },
      Object {
        "id": 3,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 4,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 5,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 6,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 7,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 8,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 9,
        "inactive": false,
        "role": null,
      },
    ],
    "periodEnd": 2017-03-26T04:42:20.000Z,
    "periodStart": 2017-02-26T05:42:20.000Z,
    "stripeId": 1,
    "stripeSubscriptionId": 1,
    "updatedAt": 2017-02-26T05:42:20.836Z,
  },
  "team": Object {
    "activeFacilitator": null,
    "facilitatorPhase": "lobby",
    "facilitatorPhaseItem": null,
    "id": 10,
    "isPaid": true,
    "meetingId": null,
    "meetingPhase": "lobby",
    "meetingPhaseItem": null,
    "orgId": 0,
    "teamName": "Team America",
  },
}
`;

exports[`addBilling changes cards for customer who had a failed charge 2`] = `
Object {
  "customers": Object {
    "account_balance": 0,
    "created": 1488087740,
    "currency": "usd",
    "default_source": "card_123",
    "delinquent": false,
    "description": null,
    "discount": null,
    "email": null,
    "id": 1,
    "livemode": false,
    "metadata": Object {
      "orgId": 0,
    },
    "object": "customer",
    "shipping": null,
    "sources": Object {
      "data": Array [
        Object {
          "address_city": null,
          "address_country": null,
          "address_line1": null,
          "address_line1_check": null,
          "address_line2": null,
          "address_state": null,
          "address_zip": null,
          "address_zip_check": null,
          "brand": "Visa",
          "country": "US",
          "customer": 1,
          "cvc_check": "pass",
          "dynamic_last4": null,
          "exp_month": "12",
          "exp_year": "2023",
          "funding": "credit",
          "id": "card_123",
          "last4": "4242",
          "metadata": Object {},
          "name": null,
          "object": "card",
          "tokenization_method": null,
        },
      ],
      "has_more": false,
      "object": "list",
      "total_count": 1,
      "url": 11,
    },
  },
  "subscriptions": Object {
    "application_fee_percent": null,
    "cancel_at_period_end": false,
    "canceled_at": null,
    "created": 1488087740,
    "current_period_end": 1490503340,
    "current_period_start": 1488087740,
    "customer": 1,
    "discount": null,
    "ended_at": null,
    "id": 1,
    "items": Object {
      "data": Array [
        Object {
          "created": 1488087740,
          "id": "si_123",
          "object": "subscription_item",
          "plan": Object {
            "amount": 19999900,
            "created": 1488087740,
            "currency": "usd",
            "id": "action-monthly-test",
            "interval": "month",
            "interval_count": 1,
            "livemode": false,
            "metadata": Object {},
            "name": "Action Monthly | TEST",
            "object": "plan",
            "statement_descriptor": null,
            "trial_period_days": null,
          },
          "quantity": 8,
        },
      ],
      "has_more": false,
      "object": "list",
      "total_count": 1,
      "url": 12,
    },
    "livemode": false,
    "metadata": Object {
      "orgId": 0,
    },
    "object": "subscription",
    "plan": Object {
      "amount": 19999900,
      "created": 1488087740,
      "currency": "usd",
      "id": "action-monthly-test",
      "interval": "month",
      "interval_count": 1,
      "livemode": false,
      "metadata": Object {},
      "name": "action-monthly",
      "object": "plan",
      "statement_descriptor": null,
      "trial_period_days": null,
    },
    "quantity": 8,
    "start": 1488087740,
    "status": "active",
    "tax_percent": 0,
    "trial_end": null,
    "trial_start": null,
  },
}
`;

exports[`addBilling changes cards for customers in good standing 1`] = `
Object {
  "organization": Object {
    "createdAt": 2017-02-26T05:42:20.836Z,
    "creditCard": Object {
      "brand": "Visa",
      "expiry": "12/23",
      "id": "card_124",
      "last4": "4242",
      "number": "4012888888881881",
    },
    "id": 0,
    "name": "The Averagers, Inc.",
    "orgUsers": Array [
      Object {
        "id": 3,
        "inactive": false,
        "role": "billingLeader",
      },
      Object {
        "id": 4,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 5,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 6,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 7,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 8,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 9,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 10,
        "inactive": false,
        "role": null,
      },
    ],
    "periodEnd": 2017-03-28T05:42:20.836Z,
    "periodStart": 2017-02-26T05:42:20.836Z,
    "stripeId": 1,
    "stripeSubscriptionId": 2,
    "updatedAt": 2017-02-26T05:42:20.836Z,
  },
}
`;

exports[`addBilling changes cards for customers in good standing 2`] = `
Object {
  "customers": Object {
    "account_balance": 0,
    "created": 1488087740,
    "currency": "usd",
    "default_source": "card_123",
    "delinquent": false,
    "description": null,
    "discount": null,
    "email": null,
    "id": 1,
    "livemode": false,
    "metadata": Object {
      "orgId": 0,
    },
    "object": "customer",
    "shipping": null,
    "sources": Object {
      "data": Array [
        Object {
          "address_city": null,
          "address_country": null,
          "address_line1": null,
          "address_line1_check": null,
          "address_line2": null,
          "address_state": null,
          "address_zip": null,
          "address_zip_check": null,
          "brand": "Visa",
          "country": "US",
          "customer": 1,
          "cvc_check": "pass",
          "dynamic_last4": null,
          "exp_month": "12",
          "exp_year": "2023",
          "funding": "credit",
          "id": "card_123",
          "last4": "4242",
          "metadata": Object {},
          "name": null,
          "object": "card",
          "tokenization_method": null,
        },
      ],
      "has_more": false,
      "object": "list",
      "total_count": 1,
      "url": 11,
    },
  },
  "subscriptions": Object {
    "application_fee_percent": null,
    "cancel_at_period_end": false,
    "canceled_at": null,
    "created": 1488087740,
    "current_period_end": 1490679740,
    "current_period_start": 1488087740,
    "customer": 1,
    "discount": null,
    "ended_at": null,
    "id": 2,
    "items": Object {
      "data": Array [
        Object {
          "created": 1488087740,
          "id": "si_123",
          "object": "subscription_item",
          "plan": Object {
            "amount": 19999900,
            "created": 1488087740,
            "currency": "usd",
            "id": "action-monthly-test",
            "interval": "month",
            "interval_count": 1,
            "livemode": false,
            "metadata": Object {},
            "name": "Action Monthly | TEST",
            "object": "plan",
            "statement_descriptor": null,
            "trial_period_days": null,
          },
          "quantity": 8,
        },
      ],
      "has_more": false,
      "object": "list",
      "total_count": 1,
      "url": 12,
    },
    "livemode": false,
    "metadata": Object {
      "orgId": 0,
    },
    "object": "subscription",
    "plan": Object {
      "amount": 19999900,
      "created": 1488087740,
      "currency": "usd",
      "id": "action-monthly-test",
      "interval": "month",
      "interval_count": 1,
      "livemode": false,
      "metadata": Object {},
      "name": "Action Monthly | TEST",
      "object": "plan",
      "statement_descriptor": null,
      "trial_period_days": null,
    },
    "quantity": 8,
    "start": 1488087740,
    "status": "active",
    "tax_percent": 0,
    "trial_end": null,
    "trial_start": null,
  },
}
`;

exports[`addBilling changes cards for customers in good standing 3`] = `
Object {
  "organization": Object {
    "createdAt": 2017-02-26T05:42:20.836Z,
    "creditCard": Object {
      "brand": "Visa",
      "expiry": "12/23",
      "id": "card_124",
      "last4": "4242",
      "number": "4012888888881881",
    },
    "id": 0,
    "name": "The Averagers, Inc.",
    "orgUsers": Array [
      Object {
        "id": 3,
        "inactive": false,
        "role": "billingLeader",
      },
      Object {
        "id": 4,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 5,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 6,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 7,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 8,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 9,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 10,
        "inactive": false,
        "role": null,
      },
    ],
    "periodEnd": 2017-03-28T05:42:20.836Z,
    "periodStart": 2017-02-26T05:42:20.836Z,
    "stripeId": 1,
    "stripeSubscriptionId": 2,
    "updatedAt": 2017-02-26T05:42:20.836Z,
  },
}
`;

exports[`addBilling changes cards for customers in good standing 4`] = `
Object {
  "customers": Object {
    "account_balance": 0,
    "created": 1488087740,
    "currency": "usd",
    "default_source": "card_123",
    "delinquent": false,
    "description": null,
    "discount": null,
    "email": null,
    "id": 1,
    "livemode": false,
    "metadata": Object {
      "orgId": 0,
    },
    "object": "customer",
    "shipping": null,
    "sources": Object {
      "data": Array [
        Object {
          "address_city": null,
          "address_country": null,
          "address_line1": null,
          "address_line1_check": null,
          "address_line2": null,
          "address_state": null,
          "address_zip": null,
          "address_zip_check": null,
          "brand": "Visa",
          "country": "US",
          "customer": 1,
          "cvc_check": "pass",
          "dynamic_last4": null,
          "exp_month": "12",
          "exp_year": "2023",
          "funding": "credit",
          "id": "card_123",
          "last4": "4242",
          "metadata": Object {},
          "name": null,
          "object": "card",
          "tokenization_method": null,
        },
      ],
      "has_more": false,
      "object": "list",
      "total_count": 1,
      "url": 11,
    },
  },
  "subscriptions": Object {
    "application_fee_percent": null,
    "cancel_at_period_end": false,
    "canceled_at": null,
    "created": 1488087740,
    "current_period_end": 1490679740,
    "current_period_start": 1488087740,
    "customer": 1,
    "discount": null,
    "ended_at": null,
    "id": 2,
    "items": Object {
      "data": Array [
        Object {
          "created": 1488087740,
          "id": "si_123",
          "object": "subscription_item",
          "plan": Object {
            "amount": 19999900,
            "created": 1488087740,
            "currency": "usd",
            "id": "action-monthly-test",
            "interval": "month",
            "interval_count": 1,
            "livemode": false,
            "metadata": Object {},
            "name": "Action Monthly | TEST",
            "object": "plan",
            "statement_descriptor": null,
            "trial_period_days": null,
          },
          "quantity": 8,
        },
      ],
      "has_more": false,
      "object": "list",
      "total_count": 1,
      "url": 12,
    },
    "livemode": false,
    "metadata": Object {
      "orgId": 0,
    },
    "object": "subscription",
    "plan": Object {
      "amount": 19999900,
      "created": 1488087740,
      "currency": "usd",
      "id": "action-monthly-test",
      "interval": "month",
      "interval_count": 1,
      "livemode": false,
      "metadata": Object {},
      "name": "Action Monthly | TEST",
      "object": "plan",
      "statement_descriptor": null,
      "trial_period_days": null,
    },
    "quantity": 8,
    "start": 1488087740,
    "status": "active",
    "tax_percent": 0,
    "trial_end": null,
    "trial_start": null,
  },
}
`;

exports[`addBilling extends trial for those still in trial period 1`] = `
Object {
  "notification": Array [],
  "organization": Object {
    "createdAt": 2017-02-26T05:42:20.836Z,
    "creditCard": Object {
      "brand": "Visa",
      "expiry": "12/23",
      "last4": "4242",
    },
    "id": 0,
    "name": "The Averagers, Inc.",
    "orgUsers": Array [
      Object {
        "id": 3,
        "inactive": false,
        "role": "billingLeader",
      },
      Object {
        "id": 4,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 5,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 6,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 7,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 8,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 9,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 10,
        "inactive": false,
        "role": null,
      },
    ],
    "periodEnd": 2017-04-27T05:42:20.000Z,
    "periodStart": 2017-02-26T05:42:20.836Z,
    "stripeId": 1,
    "stripeSubscriptionId": 2,
    "updatedAt": 2017-02-26T05:42:20.836Z,
  },
}
`;

exports[`addBilling extends trial for those still in trial period 2`] = `
Object {
  "customers": Object {
    "account_balance": 0,
    "created": 1488087740,
    "currency": "usd",
    "default_source": "card_123",
    "delinquent": false,
    "description": null,
    "discount": null,
    "email": null,
    "id": 1,
    "livemode": false,
    "metadata": Object {
      "orgId": 0,
    },
    "object": "customer",
    "shipping": null,
    "sources": Object {
      "data": Array [
        Object {
          "address_city": null,
          "address_country": null,
          "address_line1": null,
          "address_line1_check": null,
          "address_line2": null,
          "address_state": null,
          "address_zip": null,
          "address_zip_check": null,
          "brand": "Visa",
          "country": "US",
          "customer": 1,
          "cvc_check": "pass",
          "dynamic_last4": null,
          "exp_month": "12",
          "exp_year": "2023",
          "funding": "credit",
          "id": "card_123",
          "last4": "4242",
          "metadata": Object {},
          "name": null,
          "object": "card",
          "tokenization_method": null,
        },
      ],
      "has_more": false,
      "object": "list",
      "total_count": 1,
      "url": 11,
    },
  },
  "subscriptions": Object {
    "application_fee_percent": null,
    "cancel_at_period_end": false,
    "canceled_at": null,
    "created": 1488087740,
    "current_period_end": 1493271740,
    "current_period_start": 1488087740,
    "customer": 1,
    "discount": null,
    "ended_at": null,
    "id": 2,
    "items": Object {
      "data": Array [
        Object {
          "created": 1488087740,
          "id": "si_123",
          "object": "subscription_item",
          "plan": Object {
            "amount": 19999900,
            "created": 1488087740,
            "currency": "usd",
            "id": "action-monthly-test",
            "interval": "month",
            "interval_count": 1,
            "livemode": false,
            "metadata": Object {},
            "name": "Action Monthly | TEST",
            "object": "plan",
            "statement_descriptor": null,
            "trial_period_days": null,
          },
          "quantity": 8,
        },
      ],
      "has_more": false,
      "object": "list",
      "total_count": 1,
      "url": 12,
    },
    "livemode": false,
    "metadata": Object {
      "orgId": 0,
    },
    "object": "subscription",
    "plan": Object {
      "amount": 19999900,
      "created": 1488087740,
      "currency": "usd",
      "id": "action-monthly-test",
      "interval": "month",
      "interval_count": 1,
      "livemode": false,
      "metadata": Object {},
      "name": "Action Monthly | TEST",
      "object": "plan",
      "statement_descriptor": null,
      "trial_period_days": null,
    },
    "quantity": 8,
    "start": 1488087740,
    "status": "active",
    "tax_percent": 0,
    "trial_end": 1493271740,
    "trial_start": 1488087740,
  },
}
`;

exports[`addBilling starts a new subscription for those who let the trial expire 1`] = `
Object {
  "notification": Array [],
  "organization": Object {
    "createdAt": 2017-02-26T05:42:20.836Z,
    "creditCard": Object {
      "brand": "Visa",
      "expiry": "12/23",
      "last4": "4242",
    },
    "id": 0,
    "name": "The Averagers, Inc.",
    "orgUsers": Array [
      Object {
        "id": 2,
        "inactive": false,
        "role": "billingLeader",
      },
      Object {
        "id": 3,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 4,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 5,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 6,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 7,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 8,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 9,
        "inactive": false,
        "role": null,
      },
    ],
    "periodEnd": 2017-03-26T04:42:20.000Z,
    "periodStart": 2017-02-26T05:42:20.000Z,
    "stripeId": 1,
    "stripeSubscriptionId": 1,
    "updatedAt": 2017-02-26T05:42:20.836Z,
  },
  "team": Object {
    "activeFacilitator": null,
    "facilitatorPhase": "lobby",
    "facilitatorPhaseItem": null,
    "id": 10,
    "isPaid": true,
    "meetingId": null,
    "meetingPhase": "lobby",
    "meetingPhaseItem": null,
    "orgId": 0,
    "teamName": "Team America",
  },
}
`;

exports[`addBilling starts a new subscription for those who let the trial expire 2`] = `
Object {
  "customers": Object {
    "account_balance": 0,
    "created": 1488087740,
    "currency": "usd",
    "default_source": "card_123",
    "delinquent": false,
    "description": null,
    "discount": null,
    "email": null,
    "id": 1,
    "livemode": false,
    "metadata": Object {
      "orgId": 0,
    },
    "object": "customer",
    "shipping": null,
    "sources": Object {
      "data": Array [
        Object {
          "address_city": null,
          "address_country": null,
          "address_line1": null,
          "address_line1_check": null,
          "address_line2": null,
          "address_state": null,
          "address_zip": null,
          "address_zip_check": null,
          "brand": "Visa",
          "country": "US",
          "customer": 1,
          "cvc_check": "pass",
          "dynamic_last4": null,
          "exp_month": "12",
          "exp_year": "2023",
          "funding": "credit",
          "id": "card_123",
          "last4": "4242",
          "metadata": Object {},
          "name": null,
          "object": "card",
          "tokenization_method": null,
        },
      ],
      "has_more": false,
      "object": "list",
      "total_count": 1,
      "url": 11,
    },
  },
  "subscriptions": Object {
    "application_fee_percent": null,
    "cancel_at_period_end": false,
    "canceled_at": null,
    "created": 1488087740,
    "current_period_end": 1490503340,
    "current_period_start": 1488087740,
    "customer": 1,
    "discount": null,
    "ended_at": null,
    "id": 1,
    "items": Object {
      "data": Array [
        Object {
          "created": 1488087740,
          "id": "si_123",
          "object": "subscription_item",
          "plan": Object {
            "amount": 19999900,
            "created": 1488087740,
            "currency": "usd",
            "id": "action-monthly-test",
            "interval": "month",
            "interval_count": 1,
            "livemode": false,
            "metadata": Object {},
            "name": "Action Monthly | TEST",
            "object": "plan",
            "statement_descriptor": null,
            "trial_period_days": null,
          },
          "quantity": 8,
        },
      ],
      "has_more": false,
      "object": "list",
      "total_count": 1,
      "url": 12,
    },
    "livemode": false,
    "metadata": Object {
      "orgId": 0,
    },
    "object": "subscription",
    "plan": Object {
      "amount": 19999900,
      "created": 1488087740,
      "currency": "usd",
      "id": "action-monthly-test",
      "interval": "month",
      "interval_count": 1,
      "livemode": false,
      "metadata": Object {},
      "name": "action-monthly",
      "object": "plan",
      "statement_descriptor": null,
      "trial_period_days": null,
    },
    "quantity": 8,
    "start": 1488087740,
    "status": "active",
    "tax_percent": 0,
    "trial_end": 1488087740,
    "trial_start": 1488087740,
  },
}
`;

exports[`addBilling throw when the caller is not an org leader 1`] = `"{\\"_error\\":\\"Unauthorized. User is not a billing leader for that organization\\"}"`;

exports[`addBilling thrown when a bad source token is provided 1`] = `"{\\"_error\\":\\"No such token: BAD_TOKEN\\"}"`;

exports[`addBilling throws when no websocket is present 1`] = `"{\\"_error\\":\\"this must be called from a websocket\\"}"`;
